### ISSUE TITLE
CSS pseudo-selector for h1-h6?

### ISSUE THREAD

**Issue by pmaxhogan** [w3c/csswg-drafts#1008]
I frequently find myself writing CSS like this:
```CSS
h1, h2, h3, h4, h5, h6{
     ...
}
```
It annoys me that I have to type `h1, h2, h3, h4, h5, h6` so much. But it gets worse:
```CSS
article h1, article h2, article h3, article h4, article h5, article h6{
     ...
}
```
or even
```CSS
article h1.bold:hover, article h2.bold:hover, article h3.bold:hover, article h4.bold:hover, article h5.bold:hover, article h6.bold:hover{
     ...
}
```
Wouldn't it be nice to just type
```CSS
:heading{
     ...
}
```
or
```CSS
article :heading{
     ...
}
```
or 
```CSS
article :heading.bold:hover{
     ...
}
```
`:heading` just matches h1-h6. It would be easy to implement for browser vendors and easy to polyfill. But extremely helpful.

# Having a heading selector like `:heading` would be AWESOME!

**Comment by notoriousb1t** [comment]
Peanut gallery here.  :heading seems ultra specific.  Maybe we would be better served by a tag selector?  Something like 

```css
/h[0-6]/ {
   font-weight: bold;
}
```

**Comment by SelenIT** [comment]
There is the CSS Custom Selectors [proposal](https://drafts.csswg.org/css-extensions/#custom-selectors) (it even has its [PostCSS plugin](https://github.com/postcss/postcss-custom-selectors)) which would solve this exact issue and much more. But doesn't `:matches(h1, h2, h3)` from Selectors level 4 already solve the main problem here (code duplication)?

**Comment by programmer5000-com** [comment]
See #1010

**Comment by annevk** [comment]
https://github.com/whatwg/html/pull/3499 contains a `:heading`/`:heading()` pseudo-class selector proposal. @fantasai wrote that Selectors should maybe define that given that other host languages might have utility for it too. If it's successful for HTML that seems reasonable to me.

**Comment by tabatkins** [comment]
> But doesn't :matches(h1, h2, h3) from Selectors level 4 already solve the main problem here (code duplication)?

No, because HTML has ways of assigning heading levels that go beyond tagnames - `h1` nested inside of sectioning elements automatically becomes a second/third/etc level heading, as appropriate.  (Plus, `:heading` really is pretty dramatically shorter to type than `:matches(h1, h2, h3, h4, h5, h6)`, and is a pretty common thing to want to style.)

**Comment by Crissov** [comment]
Would `: heading` match any of `<caption>`, `<figcaption>`, `<label>`, `<legend>` as well?

**Comment by tabatkins** [comment]
Nah, none of those are "headings" in HTML's semantics. They're labels for some content, which serve a *similar* purpose, but aren't the same thing.

**Comment by inoas** [comment]
Is `/h[0-6]/` actually a valid selector (not to my knowledge)
But if it is not, why not make that a valid selector?

**Comment by tomhodgins** [comment]
@inoas are there any other HTML tag names, other than H1–6 where a regex-like selector like that would make sense?

If you're wanting to select tags by partial tag name like that, though CSS selectors won't currently do it, you can do this with XPath using a selector like: `//*[substring-after(name(), "h") > 0]`

Demo: https://codepen.io/tomhodgins/pen/ybgMpN

**Comment by tabatkins** [comment]
It's not a valid selector, but it's also a remarkably complex feature for something as limited as "match this family of tagnames".  LIke @tomhodgins says, there's very little call for substring-matching on tagname; insofar as we do want such a thing, prefix/suffix-matching like XPath provides probably provides 90%+ of the desired functionality.

But such a thing wouldn't even work here, because part of the point of a heading-level selector is that it works off of the HTML outline algorithm, not the tagname.

**Comment by Crissov** [comment]
Hm. Then I guess a related `:caption` selector would match the elements I mentioned and headings that are the top level in a `<section>`.

**Comment by annevk** [comment]
@emilio raised an interesting concern that I don't have a good solution for. If you have
```css
:heading(1) { ... }
:heading(2) { ... }
:heading(3) { ... }
```
and then for some reason you end up with a level 4 heading on your page, the ideal "failure mode" is that it'll render the same as 3.

Maybe instead what you want is `heading-level` as a variable that can be used by `calc()` or some such.

Thoughts?

Unless there's a relative clear answer here I'm inclined to leave this feature out of https://github.com/whatwg/html/pull/3499 and save that for a future iteration.

cc @hober

**Comment by annevk** [comment]
@js-choi's idea in https://github.com/whatwg/html/issues/3500 might be another way out here, but that looks quite involved as well.

**Comment by AmeliaBR** [comment]
> An alternative idea is saying that :heading(n) matches if the heading level is less than or equal to n.

I wouldn't want that to be the default behavior, but if this selector is adopted, I think it does make sense to add a second syntax that supports inequalities. It could be harmonized with the syntax being discussed in #4140 for `nth-child()` family of selectors:

- `heading(3)` for only level-3 headings
- `heading(n <= 3)` for level 1, 2, or 3
- `heading(n > 3)` for level 4, 5, 6, (and beyond)

**Comment by AndreasKarz** [comment]
> @inoas are there any other HTML tag names, other than H1–6 where a regex-like selector like that would make sense?
> 
> If you're wanting to select tags by partial tag name like that, though CSS selectors won't currently do it, you can do this with XPath using a selector like: `//*[substring-after(name(), "h") > 0]`
> 
> Demo: https://codepen.io/tomhodgins/pen/ybgMpN

Don't work with Selenium :-(

**Comment by a-a-GiTHuB-a-a** [comment]
%heading works…

**Comment by tabatkins** [comment]
It looks like heading-level adjustment in HTML might happen again, this time via some slightly more explicit mechanisms rather than implicitly thru the outline algorithm: <https://github.com/whatwg/html/issues/5033#issuecomment-2691535596>

So, we should step this up in priority. Anne thinks we need `:heading()` pseudo-classes alongside the HTML functionality, and I think they're right.

Just reiterating the proposal:

* `:heading` to match any heading element
* `:heading(An+B)` to match headings whose level match the An+B expression

Question: this is *effectively* a convenience selector (especially the bare `:heading`), but it also upgrades the specificity from 1 element to 1 class. Do we want to do some specificity trickery with this and have it count as an element? Or just let HTML hack around that itself for its UA stylesheet (`:is(a:not(*), :where(:heading)))` works...)? Or finally extend `:where()` to allow setting specificity explicitly, as we discussed a little bit in <https://github.com/w3c/csswg-drafts/issues/1170>?

**Comment by Crissov** [comment]
#10296 is related.

**Comment by zcorpan** [comment]
#11836 was merged.
